rumqttd/src/server/broker.rs

```diff
@@ -155,8 +156,12 @@ impl Broker {
         Ok((link_tx, link_rx))
     }
 
+    /// Starts the MQTT broker server and spawns various components like the router, cluster, metrics, and console.
+    ///
+    /// This function sets up the necessary components for the MQTT broker server to run. The function returns a
+    /// [`BrokerHandler`] that can be used to gracefully shut down the server.
     #[tracing::instrument(skip(self))]
-    pub fn start(&mut self) -> Result<(), Error> {
+    pub fn start(self) -> Result<BrokerHandler, Error> {
```

Comment:
This would make me weary about breaking current functionality. Current start function does not return, so those currently using broker start and expecting it to go forever could potentially exit their programs prematurely with this change.

Reply:
Unfortunately, when I made this change I didn't think about breaking of current API. In my case, I only need something that will stop the broker when it goes out of execution scope(tests).

I think that it would be better to control how far execution goes by a handler, rather than spawn it and then block execution till it exits. Maybe could be implemented something like that https://github.com/tokio-rs/axum/blob/main/examples/graceful-shutdown/src/main.rs

Reply:
No worries. Sorry if I'm stepping on your toes but I submitted https://github.com/bytebeamio/rumqtt/pull/929 last week so I could keep moving on my work. I think an ideal future state would be to make a new `start_async()` method for the `Broker` that would look lot what you've built here.

Reply:
No problem, I just hope that one of our PRs will be merged.